### PR TITLE
feat: [CDS-70186]: Added support for secret command step output variables

### PR DESCRIPTION
**Pull request by ivan-83**
Added support for secret command step output variables. 
Output variables defined in the Command rollback step are used during the rollback.

Having command step 1 with variables that we want to leverage in the second command step.

Screenshot 2023-06-01 at 19 52 39
Output variables are defined in YAML

```
              - step:
                  name: Deploy
                  identifier: Deploy
                  type: Command
                  timeout: 50m
                  strategy:
                    repeat:
                      items: <+stage.output.hosts>
                  spec:
                    onDelegate: false
                    environmentVariables: []
                    outputVariables:
                      - name: next_string_var
                        type: String
                        value: string_var
                      - name: next_secret_var
                        type: Secret
                        value: secret_var
                    commandUnits:
                      - identifier: script_1
                        name: script_1
                        type: Script
                        spec:
                          shell: Bash
                          source:
                            type: Inline
                            spec:
                              script: |-
                                string_var=string_var_value 
                                secret_var=secret_var_value
``` 
, and used in command step 2

Screenshot 2023-06-05 at 18 36 06

After executions, secret output variables are masked in command step 1,

Screenshot 2023-06-05 at 18 37 29

and in command step 2

Screenshot 2023-06-05 at 18 39 53

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48690)

### CODE REVIEW

**Comment by ivan-83** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/execution/sshwinrm/SshWinRmStageExecutionDetails.java [hunk 1]
```diff
@@ -45,4 +49,8 @@ public Map<String, Object> getEnvVariables() {
   public Map<String, Object> getOutVariables() {
     return outVariables;
   }
+  @Override
+  public Set<String> getSecretOutVariableNames() {
+    return Collections.emptySet();
```
Comment (vitsafronovici) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/execution/sshwinrm/SshWinRmStageExecutionDetails.java:54]: why emptySet?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java [hunk 2]
```diff
@@ -75,27 +81,83 @@ public static String getWorkingDirectory(
...
       return emptyList();
     }
 
-    List<String> outputVars = new ArrayList<>();
-    outputVariables.forEach((key, val) -> {
-      if (val instanceof ParameterField) {
-        ParameterField<?> parameterFieldValue = (ParameterField<?>) val;
-        if (parameterFieldValue.getValue() == null) {
-          throw new InvalidRequestException(format("Output variable [%s] value found to be null", key));
+    List<String> outputVarValues = new ArrayList<>();
```
Comment (acasian-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java:90]: nit `new ArrayList(outputVariables.size())`

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java [hunk 3]
```diff
@@ -75,27 +81,83 @@ public static String getWorkingDirectory(
...
+    return outputVarValues;
+  }
+
+  public static List<String> getSecretOutputVariableValues(
+      Map<String, Object> outputVariables, Set<String> secretOutputVariableNames) {
+    if (EmptyPredicate.isEmpty(outputVariables)) {
+      return emptyList();
+    }
+
+    List<String> outputVarValues = new ArrayList<>();
```
Comment (acasian-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java:114]: nit `new ArrayList(outputVariables.size())`

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/CommandTaskParameters.java [hunk 4]
```diff
@@ -55,6 +55,7 @@ public abstract class CommandTaskParameters implements TaskParameters, Execution
   SshWinRmArtifactDelegateConfig artifactDelegateConfig;
   @Expression(ALLOW_SECRETS) FileDelegateConfig fileDelegateConfig;
   @Expression(ALLOW_SECRETS) List<String> outputVariables;
+  @Expression(ALLOW_SECRETS) List<String> secretOutputVariables;
```
Comment (acasian-harness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/CommandTaskParameters.java:58]: why do w e need a separate field for secret output vars?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java [hunk 5]
```diff
@@ -75,27 +81,83 @@ public static String getWorkingDirectory(
...
+          outputVarValues.add(((ParameterField<?>) value).getValue().toString());
+        } else {
+          log.error(String.format("Not found ParameterField value for output variable [%s]. value: [%s]", name, value));
+        }
+      }
+    });
+    return outputVarValues;
+  }
+
+  public static List<String> getSecretOutputVariableValues(
```
Comment (acasian-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java:108]: i think we will need a new delegate task in order to not have kryo issues or backward compatibility.

**Review by acasian-harness — CHANGES_REQUESTED** [review]

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48690)* status: 24 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/execution/sshwinrm/SshWinRmStageExecutionDetails.java` line 54 at r1](https://reviewable.io/reviews/harness/harness-core/48690#gh-1214324647:-NXAwRQz2PP5lDKzGVa5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2a41cfff9f4ce57dd8f1111fac8fda70cc053df0/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/execution/sshwinrm/SshWinRmStageExecutionDetails.java#L54)):*
<details><summary><i>Previously, vitsafronovici (Vitalie Safronovici) wrote…</i></summary><blockquote>

why emptySet? 
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/48690#gh-1217657054:-NXAwT7268zxh8Fl0C8j:b-7l0xi8) ([raw file](https://github.com/harness/harness-core/blob/2a41cfff9f4ce57dd8f1111fac8fda70cc053df0/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java#L90)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

nit `new ArrayList(outputVariables.size())`
</blockquote></details>

The size outputVarValue list depends on two input method params outputVariables and secretOutputVariableNames. It requires additional calculations of size, hence skipping it. 
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java` line 108 at r1](https://reviewable.io/reviews/harness/harness-core/48690#gh-1217660497:-NXAxC1q3t_VMY18yCk5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2a41cfff9f4ce57dd8f1111fac8fda70cc053df0/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java#L108)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

i think we will need a new delegate task in order to not have kryo issues or backward compatibility.
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java` line 114 at r1](https://reviewable.io/reviews/harness/harness-core/48690#gh-1217657164:-NXAxEKe9WkErfkENf9G:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2a41cfff9f4ce57dd8f1111fac8fda70cc053df0/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/utils/CommandStepUtils.java#L114)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

nit `new ArrayList(outputVariables.size())`
</blockquote></details>

Done.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/CommandTaskParameters.java` line 58 at r1](https://reviewable.io/reviews/harness/harness-core/48690#gh-1217659232:-NXAxHAr2uFHg23uEqEA:bj9dw3t) ([raw file](https://github.com/harness/harness-core/blob/2a41cfff9f4ce57dd8f1111fac8fda70cc053df0/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/CommandTaskParameters.java#L58)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

why do w e need a separate field for secret output vars?
</blockquote></details>

To propagate a list of secrets from CommandStepInfo based on NGVariable. That's the only way how we for sure know the list of secrets.

**Comment by ivan-83** [comment]
trigger uts

**Review by acasian-harness — DISMISSED** [review]

**Review by vitsafronovici — DISMISSED** [review]

**Comment by ivan-83** [comment]
trigger review

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 11 of 15 files at r1, 9 of 13 files at r2, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48690)* status: 23 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/rollback/CommandStepRollbackHelper.java` line 136 at r4](https://reviewable.io/reviews/harness/harness-core/48690#-NXKGbyi9B--wvF6tbtN:-NXKGbyi9B--wvF6tbtO:b-ddignq) ([raw file](https://github.com/harness/harness-core/blob/6e36bf8fcdc7563d0a755fc1ffe5bf6b1cc5986a/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/rollback/CommandStepRollbackHelper.java#L136)):*
> ```Java
>         commandStepParameters.getOutputVariables(), commandStepParameters.getSecretOutputVariablesNames());
>     List<String> secretOutputVariables = getSecretOutputVariableValues(
>         commandStepParameters.getOutputVariables(), commandStepParameters.getSecretOutputVariablesNames());
> ```

maybe better to write 1 method which returns both "outputVariables" and "secretOutputVariables"
```
class OutputVariables {
   List<String> regularVariables;
   List<String> secretVariables;
}

OutputVariables variables = getOutputVariables(commandStepParameters.getOutputVariables(), commandStepParameters.getSecretOutputVariablesNames())
```
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/DefaultWinRmExecutor.java` line 233 at r4](https://reviewable.io/reviews/harness/harness-core/48690#-NXKHQzBBe33bu91p-AW:-NXKHQzBBe33bu91p-AX:b-7rv0o0) ([raw file](https://github.com/harness/harness-core/blob/6e36bf8fcdc7563d0a755fc1ffe5bf6b1cc5986a/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/DefaultWinRmExecutor.java#L233)):*
> ```Java
>     ShellExecutionDataBuilder executionDataBuilder = ShellExecutionData.builder();
>     secretEnvVariablesToCollect =
>         secretEnvVariablesToCollect == null ? Collections.emptyList() : secretEnvVariablesToCollect;
> ```

nit: ``` CollectionUtils.emptyIfNull(secretEnvVariablesToCollect) ```

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48690)* status: 23 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Comment by ivan-83** [comment]
trigger uts

**Review by vitsafronovici — APPROVED** [review]

**Comment by ivan-83** [comment]
trigger review

**Review by acasian-harness — APPROVED** [review]

**Pull request merged**